### PR TITLE
Fix non-standard quadpoints orders for annotations

### LINKED ISSUE #6811: Support quad points for annotations
The PDF reference states that Link and text markup annotations can specify quad points. https://www.adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/pdf_reference_1-7.pdf#page=634&zoom=auto,-246,530 explains this principle. Right now we create one container per annotation, but when quad points are defined, we need to create multiple containers with rectangles defined by a group of eight quad points. Below are known files that will benefit from quad points support, which are files where an annotation takes up multiple lines (which is not too common, fortunately):
- https://blogs.adobe.com/acrolaw/files/acrolaw/comment_print_button.pdf
- https://github.com/mozilla/pdf.js/files/37756/compressed.tracemonkey-pldi-09.pdf (from #6655)

It is also possible that the annotation's rectangle is not valid at all when quad points are specified, in which case we need to use the (at least first eight) quad points to determine the rectangle for the container. Known files are:
- http://linorg.usp.br/CTAN/macros/latex/contrib/pdfcomment/doc/pdfcomment.pdf (page 11)
- http://www.plpdf.com/examples/PLPDF-SDK/Simple_Examples/37_Markup_annotation/markup.pdf

### LINKED ISSUE #12675: Fix non-standard quadpoints orders for annotations
I have made a PDF file that serves as a minimal reduced test case for all annotation types with quadpoints support. Currently all markup annotations are supported for quadpoints (highlight, underline, strikethrough and squiggly annotations), but link annotations can also have quadpoints which we don't support yet. This can be done similar to the existing annotation types, but in my test file I actually ran into https://stackoverflow.com/questions/9855814/pdf-spec-vs-acrobat-creation-quadpoints which should be addressed first before we can do that.

In short:
- Handle https://stackoverflow.com/questions/9855814/pdf-spec-vs-acrobat-creation-quadpoints properly.
- Implement quadpoints support for link annotations.
- Include the reduced test case for all types.

### PR DESCRIPTION
**Pull request by timvandermeij**
The commit messages contain more information about the individual changes.

Fixes #12675.
Fixes #6811.

### CODE REVIEW

File: src/core/annotation.js [hunk 1]
```diff
@@ -227,7 +227,31 @@ function getQuadPoints(dict, rect) {
...
+  // mentioned on https://github.com/highkite/pdfAnnotate#QuadPoints), so
+  // this is the actual order we should work with. However, the situation is
+  // even worse since Adobe's own applications and other applications violate
+  // the specification and create annotations with other orders, namely top
+  // left, top right, bottom left and bottom right or even top left, top right,
+  // bottom right and bottom left. To avoid inconsistency and broken rendering,
+  // we normalize all lists to put the quadpoints in the same standard order
+  // (see https://stackoverflow.com/a/10729881).
+  return quadPointsLists.map(quadPointsList => {
+    const minX = Math.min(...quadPointsList.map(quadPoint => quadPoint.x));
```
Comment (calixteman) [inline src/core/annotation.js:244]: I think you could use something like
```js
quadPointsList.reduce(([mX, MX, mY, MY], p) => [Math.min(mX, p.x), Math.max(mY, p.y), Math.min(mY, p.y), Math.max(mY, p.y)]);
```
is slightly faster and more concise.
I discover few weeks ago that `reduce` is a way faster than `min+spread` (seef https://bugzilla.mozilla.org/show_bug.cgi?id=1674143)

File: src/core/annotation.js [hunk 2]
```diff
@@ -227,7 +227,31 @@ function getQuadPoints(dict, rect) {
...
+  // and top left. However, in practice PDF files use a different order,
+  // namely bottom left, bottom right, top left and top right (this is also
+  // mentioned on https://github.com/highkite/pdfAnnotate#QuadPoints), so
+  // this is the actual order we should work with. However, the situation is
+  // even worse since Adobe's own applications and other applications violate
+  // the specification and create annotations with other orders, namely top
+  // left, top right, bottom left and bottom right or even top left, top right,
+  // bottom right and bottom left. To avoid inconsistency and broken rendering,
+  // we normalize all lists to put the quadpoints in the same standard order
+  // (see https://stackoverflow.com/a/10729881).
```
Comment (Snuffleupagus) [inline src/core/annotation.js:242]: Really good comment!

However, the fact that all of this special-casing is even necessary feels somewhat sad; why is it that Annotations in particular seem to deviate *so much* from the specification in many cases :-(

**Review by Snuffleupagus — APPROVED** [review]
Looks good to me, thank you!
Also, thanks for re-factoring the existing code, since this looks much better overall :-)

File: src/core/annotation.js (hunk 1 above)
Comment (timvandermeij) [reply src/core/annotation.js]: I have changed this in the new commit. Thanks!

File: src/core/annotation.js (hunk 2 above)
Comment (timvandermeij) [reply src/core/annotation.js:242]: I feel the same. I can only guess that implementation errors were made before and fixing it properly would mean breaking existing PDF files, so they had to go with a backwards compatible solution, forcing other tools to also consider those cases. However, the fact that Adobe applications violate their own specification also gets a sad face from me :-(

**Pull request merged**